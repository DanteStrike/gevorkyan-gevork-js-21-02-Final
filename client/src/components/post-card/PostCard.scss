@import "src/assets/variables";

.post-card {
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  min-height: 300px;
  padding: 3% 5% 2% 5%;
  background-color: var(--primary-background);
  border: var(--primary-border);
  border-radius: 15px;

  &--modal {
    padding: 10px;
  }

  &--modal &__main {
    pointer-events: none;
  }

  &--modal &__img {
    min-height: 260px;
    max-height: 400px;
  }

  &--modal &__comments {
    display: block;
    width: 100%;
    background-color: var(--body-background);
    border-radius: 15px;
  }

  &--modal &__wrap {
    display: flex;
    flex-grow: 1;
    align-self: center;
    justify-content: space-between;
  }

  &--top-hide {
    min-height: 250px;
  }

  &--top-hide &__top {
    display: none;
  }

  &--top-hide &__img {
    min-height: 190px;
  }

  &__top {
    display: flex;
    align-items: flex-start;
    width: 100%;
    margin-bottom: 10px;
    border-bottom: var(--primary-border);
  }

  &__avatar {
    align-self: center;
    margin-right: 8px;
  }

  &__title {
    font-size: 16px;
    line-height: 1;
    color: var(--primary-font-color);
  }

  &__date {
    font-size: 12px;
    color: var(--secondory-font-color);
  }

  &__img {
    flex-grow: 1;
    width: 100%;
    max-height: 180px;
    margin-bottom: 5px;
    font-size: 0;
    background-color: var(--body-background);
    border-radius: 15px;
    object-fit: cover;
  }

  &__text {
    width: 100%;
    font-size: 14px;
    line-height: 1.1;
    color: var(--primary-font-color);
  }

  &__main {
    display: flex;
    flex-direction: column;
    width: 100%;
  }

  &__comments {
    display: none;
  }
}

@media (max-width: $tabletWidth) {
  .post-card {
    margin: 0 auto;
  }
}
