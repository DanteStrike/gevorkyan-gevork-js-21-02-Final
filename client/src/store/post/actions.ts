import {itemActions} from './slices/item';
import {fetchActions} from './slices/fetch';
import {listActions} from './slices/list';

const actions = {
  fetchActions,
  ...itemActions,
  ...listActions,
};

export default actions;
