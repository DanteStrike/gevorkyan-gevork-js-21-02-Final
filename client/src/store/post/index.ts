export {default as postActions} from './actions';
export {default as postOperations} from './operations';
export {default as postSelectors} from './selectors';
export {default as postReducer} from './reducers';
