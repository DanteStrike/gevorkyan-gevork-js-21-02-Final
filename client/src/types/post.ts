import {IUserPreview} from './user';

export interface IPostPreview {
  id: string;
  text: string;
  image: string;
  publishDate: string;
  owner: IUserPreview;
}
