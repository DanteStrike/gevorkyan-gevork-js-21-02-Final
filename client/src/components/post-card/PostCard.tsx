import React from 'react';
import './PostCard.scss';
import {Tooltip} from 'antd';
import {IPostPreview} from '../../types';
import LinkedAvatar from '../linked-avatar/LinkedAvatar';
import CustomLink from '../custom-link/CustomLink';
import Loading from '../loading/Loading';
import {DataUtils, RouteUtils} from '../../utils';

interface IPostCard {
  post: IPostPreview;
  hideTop?: boolean;
  modal?: boolean;
  renderComments?: () => React.ReactNode;
  isLoading?: boolean;
}

function PostCard({hideTop = false, post, modal, renderComments, isLoading = false}: IPostCard) {
  const {owner, publishDate, image, text, id} = post;
  const {firstName, lastName, title} = owner;
  const name = DataUtils.collectFullName(firstName, lastName, title);
  const profileRoute = RouteUtils.createProfileRoute(owner.id);
  const profilePostRoute = RouteUtils.createProfilePostRoute(owner.id, post.id);
  const postsPostRoute = RouteUtils.createPostsPost(post.id);

  return (
    <article className={`post-card ${hideTop ? `post-card--top-hide` : ``} ${modal ? `post-card--modal` : ``}`}>
      <Loading isLoading={isLoading} />
      <div className={`post-card__top ${hideTop ? `post-card__top--hide` : ``}`}>
        <LinkedAvatar user={owner} className="post-card__avatar" to={profileRoute} />
        <div className="post-card__wrap">
          <Tooltip
            placement="topLeft"
            title={post.id}
            getPopupContainer={() => document.querySelector(`#posts-item-${post.id}`) || document.body}
          >
            <CustomLink to={profileRoute}>
              <h2 className="post-card__title">{name}</h2>
            </CustomLink>
          </Tooltip>
          <p className="post-card__date">{publishDate}</p>
        </div>
      </div>
      <CustomLink className="post-card__main" to={hideTop ? `${profilePostRoute}` : `${postsPostRoute}`}>
        <img className="post-card__img" src={image} alt={id} />
        <p className="post-card__text">{text}</p>
      </CustomLink>
      <div className="post-card__comments">{modal && renderComments && renderComments()}</div>
    </article>
  );
}

PostCard.defaultProps = {
  hideTop: false,
  modal: false,
  renderComments: () => {},
  isLoading: false,
};

export default React.memo(PostCard);
