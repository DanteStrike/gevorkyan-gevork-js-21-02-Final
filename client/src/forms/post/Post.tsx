import React, {useEffect} from 'react';
import {useHistory} from 'react-router-dom';
import Modal from '../../components/modal/Modal';
import PostCard from '../../components/post-card/PostCard';
import useAppSelector from '../../hooks/use-app-selector';
import {postActions, postOperations, postSelectors} from '../../store/post';
import useAppDispatch from '../../hooks/use-app-dispatch';
import {RequestType} from '../../store/post/slices/fetch';
import {IComment} from '../../types';
import AppList, {AppListMod} from '../../components/app-list/AppList';
import Comment from '../../components/comment/Comment';

interface IPostProps {
  id: string;
}

function Post({id}: IPostProps) {
  const commentsPerPage = 2;

  const dispatch = useAppDispatch();
  const post = useAppSelector(postSelectors.getPost);
  const isPostLoading = useAppSelector(postSelectors.getPostIsLoading);
  const isPostError = useAppSelector(postSelectors.getPostIsError);
  const postError = useAppSelector(postSelectors.getPostError);

  const page = useAppSelector(postSelectors.getPage);
  const total = useAppSelector(postSelectors.getTotal);
  const comments = useAppSelector(postSelectors.getComments);
  const isCommentsLoading = useAppSelector(postSelectors.getCommentsIsLoading);
  const isCommentsError = useAppSelector(postSelectors.getCommentsIsError);
  const commentsError = useAppSelector(postSelectors.getCommentsError);

  const handlePaginationChange = (newPage: number) => {
    dispatch(postActions.changePage(newPage));
  };

  const history = useHistory();
  const handleModalClose = () => {
    const {pathname} = history.location;
    const newRout = pathname.substring(0, pathname.lastIndexOf('/'));
    history.push(newRout);
  };

  useEffect(() => {
    dispatch(postOperations.loadPost(id));
  }, [dispatch, id]);

  useEffect(() => {
    dispatch(postOperations.loadComments(id, commentsPerPage, page));
  }, [dispatch, id, page]);

  useEffect(
    () => () => {
      dispatch(postActions.fetchActions[RequestType.LOAD_POST].requestAbort());
      dispatch(postActions.fetchActions[RequestType.LOAD_COMMENTS].requestAbort());
      dispatch(postActions.reset());
      dispatch(postActions.resetList());
    },
    [dispatch]
  );

  const renderComments = () => (
    <AppList
      current={page}
      total={total}
      dataSource={comments}
      onChange={handlePaginationChange}
      isLoading={isCommentsLoading}
      pageSize={commentsPerPage}
      renderItem={(comment: IComment) => (
        <AppList.Item key={comment.id}>
          <Comment comment={comment} />
        </AppList.Item>
      )}
      mod={AppListMod.SMALL_ROW}
    />
  );

  return (
    <Modal
      isOpen
      onClose={handleModalClose}
      isError={isPostError || isCommentsError}
      errMsg={postError || commentsError}
    >
      <PostCard post={post} modal renderComments={renderComments} isLoading={isPostLoading} />
    </Modal>
  );
}

export default Post;
