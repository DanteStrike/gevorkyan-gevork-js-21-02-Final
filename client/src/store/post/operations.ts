import {AxiosInstance} from 'axios';
import actions from './actions';
import i18next from '../../locale/i18next';
import {IComments, IPostPreview} from '../../types';
import {FetchErrorType} from '../../enums';
import {RequestType} from './slices/fetch';

const loadPost = (id: string) => (dispatch: any, getState: any, api: AxiosInstance) => {
  const controller = new AbortController();

  api
    .get<IPostPreview>(`/post/${id}?locale=${i18next.resolvedLanguage}`, {signal: controller.signal})
    .then((response) => {
      dispatch(actions.set(response.data));
      dispatch(actions.fetchActions[RequestType.LOAD_POST].requestFinished());
    })
    .catch((err) => {
      if (err.response?.data?.error === FetchErrorType.PARAMS_NOT_VALID) {
        dispatch(actions.fetchActions[RequestType.LOAD_POST].requestFailed(i18next.t(`error.network.notFoundPost`)));
      } else {
        dispatch(actions.fetchActions[RequestType.LOAD_POST].requestFailed(err.message));
      }
    });
  dispatch(actions.fetchActions[RequestType.LOAD_POST].requestStart(controller));
};

const loadComments =
  (id: string, limit: number, page: number) => (dispatch: any, getState: any, api: AxiosInstance) => {
    const controller = new AbortController();

    api
      .get<IComments>(`/post/${id}/comment?page=${page - 1}&limit=${limit}&locale=${i18next.resolvedLanguage}`, {
        signal: controller.signal,
      })
      .then((response) => {
        dispatch(actions.setup(response.data));
        dispatch(actions.fetchActions[RequestType.LOAD_COMMENTS].requestFinished());
      })
      .catch((err) => {
        if (err.response?.data?.error === FetchErrorType.PARAMS_NOT_VALID) {
          dispatch(
            actions.fetchActions[RequestType.LOAD_COMMENTS].requestFailed(i18next.t(`error.network.notFoundPost`))
          );
        } else {
          dispatch(actions.fetchActions[RequestType.LOAD_COMMENTS].requestFailed(err.message));
        }
      });
    dispatch(actions.fetchActions[RequestType.LOAD_COMMENTS].requestStart(controller));
  };

export default {
  loadPost,
  loadComments,
};
