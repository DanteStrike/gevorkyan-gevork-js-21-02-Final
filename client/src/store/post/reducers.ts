import {combineReducers} from 'redux';
import {fetchReducer, RequestType} from './slices/fetch';
import {itemReducer} from './slices/item';
import NameSpace from '../name-space';
import {IFetchStore, IListStore} from '../hor';
import {IComment, IPostPreview} from '../../types';
import {listReducer} from './slices/list';

const post = itemReducer;
const reducer = combineReducers({
  fetch: fetchReducer,
  post,
  comments: listReducer,
});
export interface IPostDuckStore {
  [NameSpace.POST]: {
    fetch: {
      [RequestType.LOAD_POST]: IFetchStore;
      [RequestType.LOAD_COMMENTS]: IFetchStore;
    };
    post: IPostPreview;
    comments: IListStore<IComment>;
  };
}
export default reducer;
