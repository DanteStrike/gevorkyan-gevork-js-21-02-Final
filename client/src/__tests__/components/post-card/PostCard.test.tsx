import {shallow} from 'enzyme';
import React from 'react';
import PostCard from '../../../components/post-card/PostCard';
import {postOne} from '../../mocks/post.mock';

jest.mock(`../../../utils/route`, () => ({
  __esModule: true,
  default: {
    ...jest.requireActual(`../../../utils/route`).default,
    createProfilePostRoute: () => `profileRoute`,
    createPostsPost: () => `postRoute`,
  },
}));

describe(`Component RegForm should render correctly`, () => {
  it(`snapshot default test`, () => {
    const wrapper = shallow(<PostCard post={postOne} />);
    expect(wrapper).toMatchSnapshot();
  });

  it(`snapshot loading test`, () => {
    const wrapper = shallow(<PostCard post={postOne} isLoading />);
    expect(wrapper).toMatchSnapshot();
  });

  it(`snapshot comments render test`, () => {
    const wrapper = shallow(<PostCard post={postOne} modal renderComments={() => <div>Comments</div>} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe(`Component RegForm should work correctly`, () => {
  it(`card modes test`, () => {
    const wrapper = shallow(<PostCard post={postOne} />);

    expect(wrapper.find(`article.post-card--top-hide`)).toHaveLength(0);
    expect(wrapper.find(`.post-card__top.post-card__top--hide`)).toHaveLength(0);
    expect(wrapper.find(`article.post-card--modal`)).toHaveLength(0);
    expect(wrapper.find(`.post-card__main`).prop(`to`)).toBe(`postRoute`);

    wrapper.setProps({hideTop: true});
    wrapper.update();

    expect(wrapper.find(`article.post-card--top-hide`)).toHaveLength(1);
    expect(wrapper.find(`.post-card__top.post-card__top--hide`)).toHaveLength(1);
    expect(wrapper.find(`article.post-card--modal`)).toHaveLength(0);
    expect(wrapper.find(`.post-card__main`).prop(`to`)).toBe(`profileRoute`);

    wrapper.setProps({modal: true});
    wrapper.update();

    expect(wrapper.find(`article.post-card--top-hide`)).toHaveLength(1);
    expect(wrapper.find(`.post-card__top.post-card__top--hide`)).toHaveLength(1);
    expect(wrapper.find(`article.post-card--modal`)).toHaveLength(1);
    expect(wrapper.find(`.post-card__main`).prop(`to`)).toBe(`profileRoute`);
  });
});
