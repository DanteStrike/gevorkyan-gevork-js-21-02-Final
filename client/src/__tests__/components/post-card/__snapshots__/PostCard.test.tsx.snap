// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component RegForm should render correctly snapshot comments render test 1`] = `
<article
  className="post-card  post-card--modal"
>
  <Memo(Loading)
    isLoading={false}
  />
  <div
    className="post-card__top "
  >
    <Memo(LinkedAvatar)
      className="post-card__avatar"
      to="/profile/userID_1"
      user={
        Object {
          "firstName": "anyFN",
          "id": "userID_1",
          "lastName": "antLN",
          "picture": "url",
          "title": "ms",
        }
      }
    />
    <div
      className="post-card__wrap"
    >
      <Tooltip
        arrowPointAtCenter={false}
        autoAdjustOverflow={true}
        getPopupContainer={[Function]}
        mouseEnterDelay={0.1}
        mouseLeaveDelay={0.1}
        placement="topLeft"
        title="postID_1"
      >
        <Memo(CustomLink)
          to="/profile/userID_1"
        >
          <h2
            className="post-card__title"
          >
            Ms. antLN anyFN
          </h2>
        </Memo(CustomLink)>
      </Tooltip>
      <p
        className="post-card__date"
      >
        un-normalized
      </p>
    </div>
  </div>
  <Memo(CustomLink)
    className="post-card__main"
    to="postRoute"
  >
    <img
      alt="postID_1"
      className="post-card__img"
      src="postIMG_URL"
    />
    <p
      className="post-card__text"
    >
      postText
    </p>
  </Memo(CustomLink)>
  <div
    className="post-card__comments"
  >
    <div>
      Comments
    </div>
  </div>
</article>
`;

exports[`Component RegForm should render correctly snapshot default test 1`] = `
<article
  className="post-card  "
>
  <Memo(Loading)
    isLoading={false}
  />
  <div
    className="post-card__top "
  >
    <Memo(LinkedAvatar)
      className="post-card__avatar"
      to="/profile/userID_1"
      user={
        Object {
          "firstName": "anyFN",
          "id": "userID_1",
          "lastName": "antLN",
          "picture": "url",
          "title": "ms",
        }
      }
    />
    <div
      className="post-card__wrap"
    >
      <Tooltip
        arrowPointAtCenter={false}
        autoAdjustOverflow={true}
        getPopupContainer={[Function]}
        mouseEnterDelay={0.1}
        mouseLeaveDelay={0.1}
        placement="topLeft"
        title="postID_1"
      >
        <Memo(CustomLink)
          to="/profile/userID_1"
        >
          <h2
            className="post-card__title"
          >
            Ms. antLN anyFN
          </h2>
        </Memo(CustomLink)>
      </Tooltip>
      <p
        className="post-card__date"
      >
        un-normalized
      </p>
    </div>
  </div>
  <Memo(CustomLink)
    className="post-card__main"
    to="postRoute"
  >
    <img
      alt="postID_1"
      className="post-card__img"
      src="postIMG_URL"
    />
    <p
      className="post-card__text"
    >
      postText
    </p>
  </Memo(CustomLink)>
  <div
    className="post-card__comments"
  />
</article>
`;

exports[`Component RegForm should render correctly snapshot loading test 1`] = `
<article
  className="post-card  "
>
  <Memo(Loading)
    isLoading={true}
  />
  <div
    className="post-card__top "
  >
    <Memo(LinkedAvatar)
      className="post-card__avatar"
      to="/profile/userID_1"
      user={
        Object {
          "firstName": "anyFN",
          "id": "userID_1",
          "lastName": "antLN",
          "picture": "url",
          "title": "ms",
        }
      }
    />
    <div
      className="post-card__wrap"
    >
      <Tooltip
        arrowPointAtCenter={false}
        autoAdjustOverflow={true}
        getPopupContainer={[Function]}
        mouseEnterDelay={0.1}
        mouseLeaveDelay={0.1}
        placement="topLeft"
        title="postID_1"
      >
        <Memo(CustomLink)
          to="/profile/userID_1"
        >
          <h2
            className="post-card__title"
          >
            Ms. antLN anyFN
          </h2>
        </Memo(CustomLink)>
      </Tooltip>
      <p
        className="post-card__date"
      >
        un-normalized
      </p>
    </div>
  </div>
  <Memo(CustomLink)
    className="post-card__main"
    to="postRoute"
  >
    <img
      alt="postID_1"
      className="post-card__img"
      src="postIMG_URL"
    />
    <p
      className="post-card__text"
    >
      postText
    </p>
  </Memo(CustomLink)>
  <div
    className="post-card__comments"
  />
</article>
`;
